### PR TITLE
Feature: reorganize ManagedTransport concurrency

### PR DESCRIPTION
**Pull request by ayuryshev**
There are 7 changes in this PR.

This PR includes changes in PR #380  - that solved issue with failing first message.

### 2. `Manager.entries` now unique and `ReconnectTransports` now thread-safe

Changes:
    
1. Manager.entries now map[Entry]struct{} (was slice)
2. Work with with them in ReconnectTransports now locked by mutex
    properly
    
Tested both in skywire, skywire-services and by integration tests

### 3. `router.Serve`, `setup.Node`: serving accepted and dialled transports in the same loop

1. `ManagedTransport`: new field `Accepted bool` - to distinguish dialled
and accepted transports
2. `Manager`: it's only one channel now for both accepted and dialled
transports
3. `Router.Serve`: it's only one loop for both accepted and dialled
transports
4. `setup.Node` changes in the same way
   
### 4. More strict typing in PacketRouter, Router, mockRouter

Removed overgeneralization of `ManagedTransport` as `Transport` where it makes no sense.

### 5. Manager.manageTransport

Goroutines in Manager.acceptTransport and Manager.createTransport are    extracted into method  Manager.manageTransport

### 7. Incorrect shutdown: redialling on close

Solved.

Closes: #327
Relates to: #364

### CODE REVIEW

File: pkg/transport/manager.go [hunk 1]
```diff
@@ -259,31 +255,49 @@ func (tm *Manager) DeleteTransport(id uuid.UUID) error {
...
 		tr.Close()
 	}
-	tm.transports = make(map[uuid.UUID]*ManagedTransport)
 	tm.mu.Unlock()
 
 	if _, err := tm.config.DiscoveryClient.UpdateStatuses(context.Background(), statuses...); err != nil {
 		tm.Logger.Warnf("Failed to change transport status: %s", err)
 	}
 
+	for _, f := range tm.factories {
```
Comment (ayuryshev) [inline pkg/transport/manager.go:277]: This was the first cause of redialling  during shutdown

File: pkg/transport/manager.go [hunk 2]
```diff
@@ -416,10 +379,43 @@ func (tm *Manager) walkEntries(walkFunc func(*Entry) bool) *Entry {
...
 }
 
+func (tm *Manager) manageTransport(ctx context.Context, managedTr *ManagedTransport, factory Factory, remote cipher.PubKey, public bool, accepted bool) {
+	select {
+	case <-managedTr.doneChan:
+		tm.Logger.Infof("Transport %s closed", managedTr.ID)
+		return
+	case err := <-managedTr.errChan:
+		managedTr.mu.Lock()
+		isClosing := managedTr.isClosing
```
Comment (ayuryshev) [inline pkg/transport/manager.go]: And that prevents redialling on errors during shutdown

File: pkg/transport/manager.go [hunk 3]
```diff
@@ -368,33 +346,18 @@ func (tm *Manager) acceptTransport(ctx context.Context, factory Factory) (*Manag
...
 	}
 	tm.mu.Unlock()
 
-	// go func(managedTr *ManagedTransport, tm *Manager) {
-	go func() {
-		select {
-		case <-managedTr.doneChan:
-			tm.Logger.Infof("Transport %s closed", managedTr.ID)
-			return
-		case <-tm.doneChan:
```
Comment (ayuryshev) [inline pkg/transport/manager.go:388]: And this was very subtle bug. Must be more evident with messaging tests with more than 2 skychat-nodes

**Review by evanlinjin — APPROVED** [review]

File: pkg/transport/manager.go [hunk 4]
```diff
@@ -54,23 +53,18 @@ func NewManager(config *ManagerConfig, factories ...Factory) (*Manager, error) {
...
-		entries:        mEntries,
-		AcceptedTrChan: make(chan *ManagedTransport, 10),
-		DialedTrChan:   make(chan *ManagedTransport, 10),
-		doneChan:       make(chan struct{}),
+		Logger:     logging.MustGetLogger("trmanager"),
+		config:     config,
+		factories:  fMap,
+		transports: make(map[uuid.UUID]*ManagedTransport),
+		entries:    mEntries,
+		// AcceptedTrChan: make(chan *ManagedTransport, 10),
```
Comment (ivcosla) [inline pkg/transport/manager.go]: Remove this before merging

**Review by ivcosla — COMMENTED** [review]
Is looking good for me, however please wait for integration test input before merging it.

File: pkg/transport/manager.go (hunk 4 above)
Comment (ayuryshev) [reply pkg/transport/manager.go]: Done

**Comment by ivcosla** [comment]
Overall is looking better than mainnet. Chat app restart for itnermediary node is quite reliable.

At first I had trouble running the PR because after the first run I restarted nodes and they all hanged on startup. Cleaning redis didn't help, the problem had to do with routing.db, however have not been able to reproduce this error again.

SSH is still working well, though we still have the problem where server and client apps hang after node shutdown. There is some problem with the integration test for SSH parsing arguments (cannot parse `echo`), but this is also present in mainnet at the moment, guess it was introduced during renaming.

Very good job!

**Review by ivcosla — APPROVED** [review]

**Pull request merged**